internal/sim/sim.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sim
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"reflect"
+	"sync"
+
+	"github.com/ServiceWeaver/weaver/internal/config"
+	"github.com/ServiceWeaver/weaver/internal/weaver"
+	"github.com/ServiceWeaver/weaver/runtime"
+	"github.com/ServiceWeaver/weaver/runtime/codegen"
+	"github.com/ServiceWeaver/weaver/runtime/protos"
+	"golang.org/x/exp/slog"
+)
+
+type Options struct {
+	Seed           int64
+	MaxReplicas    int
+	ConfigFilename string
+	Config         string
+}
+
+type Simulator struct {
```

Comment:
Is this thing meant to implement the `Weavelet` interface? If so, should we rename it to `SimulatorWeavelet`?

(I suppose that if you you mean to drive it using `weaver.Run`, then you'd have to make it implement the `Weavelet` interface.)

Reply:
It implements `GetIntf` but not `GetImpl`, so it doesn't implement `Weavelet` for now. I might remove (or unexport) `GetIntf` in the future as well. Not sure yet. I don't plan on using `weaver.Run` though!